crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -411,42 +476,176 @@ impl Object {
+            K::WrappedOrDeleted(_) | K::OutsideAvailableRange => None,
+        }
+    }
+
+    pub(crate) fn version_impl(&self) -> Option<u64> {
+        use ObjectKind as K;
+
+        match &self.kind {
+            K::Live(native, _) | K::NotIndexed(native) | K::Historical(native, _) => {
+                Some(native.version().value())
+            }
+            K::WrappedOrDeleted(stored) => Some(stored.object_version as u64),
+            K::OutsideAvailableRange => None,
+        }
+    }
+
+    async fn query_live(db: &Db, address: SuiAddress) -> Result<Option<Self>, Error> {
+        let vec_address = address.into_vec();
+        use objects::dsl as objects;
+
+        let stored_obj: Option<StoredObject> = db
+            .execute(move |conn| {
+                conn.result(move || {
+                    objects::objects
+                        .filter(objects::object_id.eq(vec_address.clone()))
+                        .limit(1)
+                        .into_boxed()
+                })
+                .optional()
+            })
+            .await
+            .map_err(|e| Error::Internal(format!("Failed to fetch object: {e}")))?;
+
+        stored_obj.map(Self::try_from).transpose()
+    }
+
+    async fn query_historical(
         db: &Db,
         address: SuiAddress,
         version: Option<u64>,
+        checkpoint_sequence_number: Option<u64>,
     ) -> Result<Option<Self>, Error> {
-        use objects::dsl;
+        let vec_address = address.into_vec();
+        let version = version.map(|n| n as i64);
+        let checkpoint_sequence_number = checkpoint_sequence_number.map(|n| n as i64);
 
-        let address = address.into_vec();
-        let version = version.map(|v| v as i64);
+        use checkpoints::dsl as checkpoints;
+        use objects_history::dsl as history;
+        use objects_snapshot::dsl as snapshot;
 
-        let stored_obj: Option<StoredObject> = db
+        let results: Option<Vec<StoredHistoryObject>> = db
             .execute(move |conn| {
-                conn.first(move || {
-                    let mut query = dsl::objects
-                        .filter(dsl::object_id.eq(address.clone()))
+                conn.results(move || {
+                    // If an object was created or mutated in a checkpoint outside the current
+                    // available range, and never touched again, it will not show up in the
+                    // objects_history table. Thus, we always need to check the objects_snapshot
+                    // table as well.
+                    let mut snapshot_query = snapshot::objects_snapshot
+                        .filter(snapshot::object_id.eq(vec_address.clone()))
+                        .into_boxed();
+
+                    let mut historical_query = history::objects_history
+                        .filter(history::object_id.eq(vec_address.clone()))
+                        .order_by(history::object_version.desc())
```

Comment:
a couple of qs:
- `objects_history` may contains object versions later than `checkpoint_sequence_number`, should we filter here with `<= cp_seq_number`?
- here by limit 1, we only pick the latest, is it possible that a version in (snapshot obj version, latest obj version in object_history) is desired instead?